Add text icons for HQ items and collectables

augenfrosch opened the pull request:
Adds text icons for HQ items and collectables using the same Unicode characters Final Fantasy XIV uses.
Additionally, improves the handling of pasted text copied from the game and changes `game_data::Item` to use `AlwaysCollectable` instead of `IsCollectable` from `Item.csv`.

**Overview of the changes:**
- Add text icons for HQ items and collectables
  - Add a font that recreates the icons
  - Use the HQ character `'\u{e03c}'` instead of "(HQ)" when displaying the name of HQ items
  - Add the collectable character `'\u{e03d}'` to the end of the name of collectables
  - User input copied from the game that includes these characters now displays correctly
- Improve handling of pasted text copied from the game
  - Filter out any null bytes that are included in the search text
  - Remove the HQ character from the search
- Change `game_data::Item` to use `AlwaysCollectable` instead of `IsCollectable` from `Item.csv`.

**Caveats / Limitations caused by the current implementation:**
- The font has a relatively "large" file size even though it just adds two characters. I couldn't get it any smaller than ~2kB with my limited knowledge
- The code that checks if the collectable character should be added is used across all items, as such, it is also run for food items and potions.
  - It would be possible to bake the character into the item name but that would make the field differ from the entry in `Item.csv` which is potentially undesirable. Alternatively, you could pass an additional argument (`can_be_collectable`) to `game_data::get_item_name`
- Null bytes are filtered from the search text regardless of the target. Although I only observed the issue on a native Windows build, I assume it happens when no code that affects the clipboard uses null-terminated strings. As such, I kept the code target independent
- Removing the HQ character from the food and potion search makes the search less specific but I personally found it useful to always show both the NQ and HQ version.
  - The current version of Raphael allows for the specification by searching for "[Food/Potion] (HQ)" which only matches the HQ version of the item. If this behaviour is desired, I could change it to not remove the HQ character for the search to restore the old behaviour
- Related to the last point, there is currently no easy way to include the HQ and collectable characters in the search text without having them in the clipboard
  - It would be possible to replace "(HQ)" and something like "(C)" with the respective character

**Screenshot:**
![image](https://github.com/user-attachments/assets/adec7ef1-f78b-4933-a3b6-4c23efbe0928)

KonaeAkira commented on src/widgets/food_select.rs:
```diff
@@ -83,14 +83,18 @@ impl<'a> Widget for FoodSelect<'a> {
                     ui.label("Search:");
-                    ui.text_edit_singleline(&mut search_text);
+                    if ui.text_edit_singleline(&mut search_text).changed() {
+                        search_text.truncate(search_text.trim_end_matches('\0').len());
```
If `trim_end_matches` trims from the beginning of the string, then this line is wrong. Although that case will probably never occur, it is better to rewrite this as
```rust
search_text = search_text.trim_end_matches('\0').to_string();
```

KonaeAkira replied:
Ah I just realized that `trim_end_matches` only trims from the end :smiley:
Still, I prefer the suggested version for readability.

KonaeAkira replied:
Even better, you could use
```rust
search_text.remove_matches('/0');
```

augenfrosch replied:
I used `trim_end_matches` since the null bytes are only at the end of the string when pasting from the game and `replace` would go through all characters instead of just the ones at the end until one doesn't match.
The use of `truncate` is because I wanted to make absolutely sure that no new `String` is created, maybe it is not needed.
But I can definitely change it.

KonaeAkira replied:
The performance hit from creating a new string is negligible, since this part is only execute once after the input changes. That's why I would prefer simplicity and readability over performance in this case.

KonaeAkira commented on src/widgets/potion_select.rs:
```diff
@@ -83,14 +83,18 @@ impl<'a> Widget for PotionSelect<'a> {
                     ui.label("Search:");
-                    ui.text_edit_singleline(&mut search_text);
+                    if ui.text_edit_singleline(&mut search_text).changed() {
+                        search_text.truncate(search_text.trim_end_matches('\0').len());
```
Same as previous comment. Change this to
```rust
search_text = search_text.trim_end_matches('\0').to_string();
```

KonaeAkira commented on src/widgets/recipe_select.rs:
```diff
@@ -68,14 +68,17 @@ impl<'a> RecipeSelect<'a> {
             ui.label("Search:");
-            ui.text_edit_singleline(&mut search_text);
+            if ui.text_edit_singleline(&mut search_text).changed() {
+                search_text.truncate(search_text.trim_end_matches('\0').len());
```
Same as previous comment. Change this to
```rust
search_text = search_text.trim_end_matches('\0').to_string();
```

augenfrosch commented:
I made the requested change and (hopefully) fixed the failing test. One of the tests (`test_item_name_44232_hq`) had to be rewritten since with the current implementation an item that is always a collectable will only include the collectable character at the end of it's name. To test the HQ functionality, I added a new test (`test_item_name_44104_hq`) that replaces the role of the aforementioned test 

Additionally, I implemented two of my proposed alterations that I would now consider a usability improvement:
- The search for food items and potions now behaves the same as the current version of Raphael, by not excluding the HQ character.
- I added the easy input method for the special characters

Without these changes a user can not easily search for the same things possible in the current version of Raphael, specifically searches that included "(HQ)" at the end were not possible.

KonaeAkira commented:
Nice work! Thank you for your contribution :)
